app/logs/zerolog-logger.go

```diff
@@ -0,0 +1,135 @@
+// Package logs provides a global logger for zerolog.
+// Based on github.com/rs/zerolog/log of Olivier Poitrey
+package logs
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/rs/zerolog"
+)
+
+// Output duplicates the global logger and sets w as its output.
+func Output(w io.Writer) zerolog.Logger {
```

Comment:
app/logs is centrally configured structure multi-logger
app/logs/zerolog-logger.go contains helper functions for github.com/rs/zerolog; and they already use initialised configured logger with caller/stack info logging

The func Output(w io.Writer) zerolog.Logger allows to add one more output for example to a file. It will work in place of usage only. While app/logs can be configured in one place for all usages
This helper func Output is not used in code but it is here to keep 'standard' of zerolog

Reply:
in fact Output can use any io.Writer
In past I had set logging directly to Splunk HTTP Event Collector which used to extract structured logged fields much better then Splunk parsing syslog for the same structures
It can use io.Writer sending log messages to db too ...

Reply:
My question is, why not directly call `github.com/rs/zerolog` `log2C.Output` ? It has the exact same type signature as your `Output()` function.

And on top of that, by creating `LogC2` initialized logger, you can never write tests for logs - you can never swap what you initialized as logger with different logger because it is global to this package.

What I mean by testing logs - for example, we have audit logging in some of our code, where the business requirement is specifically to log certain data after an event. To test your business logic, you have to call the event handling function with a log observer, where you can test for it containing certain data.

Reply:
app/logs allows central, flexible configuration for all logs
The app/logs is not yet fully integrated because it uses config out of app config and it can be the main source of confusion for now
The intention is to have configurable and consistent behaviour of logs in instance of app. No matter if someone  try to use standard log or app.logs both will go to the same outputs and will have predefined fields entered

To solve your audit logging by app.logs there is required to define one more output  (similar as it was done for Splunk HEC). Then depends if all events should go to this output=observer then it should be added to Logger or if only particular events then we need additional LogX with added output=observer and to use logs.LogX in these particular places (if LogX will be based on Logger then other outputs also receive it's events if not than sending to other will require two loggers in particular places)

app/logs was initially created to solve issue with three separate in-house loggers in one codebase to format and direct all of them to configured outputs (depends on env) and with requirement to not change code in place of usage.  It was really tricky. In comparison your audit logging seams to be very logic and straight to implement.

In app/logs there is preconfigured github.com/rs/zerolog so global defaults of rs/zerolog are changed 
```go
	zerolog.TimeFieldFormat = zerolog.TimeFormatUnixMicro
	zerolog.TimestampFieldName = "µs"
	zerolog.ErrorStackMarshaler = pkgerrors.MarshalStack //nolint:reassign
	zerolog.CallerMarshalFunc = formatCaller
	// zerolog.SetGlobalLevel(zerolog.Disabled)
	// zerolog.SetGlobalLevel(zerolog.DebugLevel)
```
and there are added output(s) and some fields to default Logger
```go
	Logger = zerolog.New(zerolog.MultiLevelWriter(
		c.setSyslog(),
		c.setStdErr(),
		c.setStdOut(),
		c.setConsole(),
	)).
		With().
		Timestamp().
		Str("app", c.App).
		Int("pid", os.Getpid()).
		Logger()
```
and there are separate loggers with different caller
```go
	LogC2 = Logger.With().
		CallerWithSkipFrameCount(2).
```
```go
	LogC3 = Logger.With().
		CallerWithSkipFrameCount(3).
```
and standard log is configured to use Logger output(s) and correct stack caller
```go
	log.SetFlags(0)
	log.SetOutput(Logger.With().
		CallerWithSkipFrameCount(5).
```

Then logs.Output() already uses preconfigured outputs and fields as LogC2
It allows to add additional output (if required, but not used in current app)
```go
func Output(w io.Writer) zerolog.Logger {
	return LogC2.Output(w)
```

While zerolog.Output() will add output to global defaults of rs/zerolog
```go
var Logger = zerolog.New(os.Stderr).With().Timestamp().Logger()

func Output(w io.Writer) zerolog.Logger {
	return Logger.Output(w)
```